clients/drcachesim/tools/common/decode_cache.h

```diff
@@ -0,0 +1,469 @@
+ * heavy lifting of actually producing the decoded #instr_t. The decoded #instr_t
+ * may be made to persist beyond the set_decode_info() calls by constructing the
+ * #decode_cache_t object with \p persist_decoded_instrs_ set to true.
+ *
+ * Usage note: after constructing an object, init() must be called.
+ */
+template <class DecodeInfo> class decode_cache_t : public decode_cache_base_t {
+    static_assert(std::is_base_of<decode_info_base_t, DecodeInfo>::value,
+                  "DecodeInfo not derived from decode_info_base_t");
+
+public:
+    decode_cache_t(void *dcontext, bool persist_decoded_instrs)
+        : dcontext_(dcontext)
+        , persist_decoded_instrs_(persist_decoded_instrs) {};
+    virtual ~decode_cache_t()
+    {
+    }
+
+    /**
+     * Returns a pointer to the DecodeInfo available for the instruction at \p pc.
+     * Returns nullptr if no instruction is known at that \p pc. Returns the
+     * default-constructed DecodeInfo if an instr was seen at that \p pc but there
+     * was a decoding error for the instruction.
+     *
+     * Guaranteed to be non-nullptr and valid if the prior add_decode_info() for
+     * that \p pc returned no error.
+     *
+     * When analyzing #memref_t from a trace, it may be better to just use
+     * add_decode_info() instead (as it also returns the added DecodeInfo) if
+     * it's possible that the instr at \p pc may have changed (e.g., JIT code).
+     */
+    DecodeInfo *
+    get_decode_info(app_pc pc)
+    {
+        auto it = decode_cache_.find(pc);
+        if (it == decode_cache_.end()) {
+            return nullptr;
+        }
+        return &it->second;
+    }
+    /**
+     * Adds decode info for the given \p memref_instr if it is not yet recorded.
+     * or if it contains a new encoding.
+     *
+     * Uses the embedded encodings in the trace or, if init() was invoked with
+     * a module file path, the encodings from the instantiated
+     * #module_mapper_t.
+     *
+     * If there is a decoding failure, the default-constructed DecodeInfo that
+     * returns is_valid() = false will be added to the cache.
+     *
+     * Returns a pointer to whatever DecodeInfo is present in the cache in the
+     * \p cached_decode_info reference pointer parameter, or a nullptr if none
+     * is cached. This helps avoid a repeated lookup in a subsequent
+     * get_decode_info() call.
+     *
+     * Returns the error string, or an empty string if there was no error. A
+     * valid DecodeInfo is guaranteed to be added to the cache if there was no
+     * error.
+     */
+    std::string
+    add_decode_info(const dynamorio::drmemtrace::_memref_instr_t &memref_instr,
+                    DecodeInfo *&cached_decode_info)
+    {
+        cached_decode_info = nullptr;
+        if (!init_done_) {
+            return "init() must be called first";
+        }
+        const app_pc trace_pc = reinterpret_cast<app_pc>(memref_instr.addr);
+        auto it_inserted = decode_cache_.emplace(
```

Comment:
Add a XXX comment: once we have C++17 we'll use try_emplace and avoid the creation when it already exists.

For C++11: you're assuming the creation is not costly and is better than find()?  The common case is it already existing: for long traces it should already exist nearly every time.  Is emplace() really better than find() then?

Reply:
Yes, try_emplace is what we eventually want, and I'll the comment.

> For C++11: you're assuming the creation is not costly and is better than find()? 

Yes: sounded reasonable to me initially. But maybe it is better to do an explicit .find to emulate try_emplace.

Reply:
Done